storm-core/src/jvm/org/apache/storm/scheduler/IsolationScheduler.java

```diff
@@ -0,0 +1,434 @@
+                ids.add(topology.getId());
+            }
+        }
+        return ids;
+    }
+
+    // map from topology id -> set of sets of executors
+    private Map<String, Set<Set<ExecutorDetails>>> topologyWorkerSpecs(List<TopologyDetails> topologies) {
+        Map<String, Set<Set<ExecutorDetails>>> workerSpecs = new HashMap<String, Set<Set<ExecutorDetails>>>();
+        for (TopologyDetails topology : topologies) {
+            workerSpecs.put(topology.getId(), computeWorkerSpecs(topology));
+        }
+        return workerSpecs;
+    }
+
+    private Map<String, List<AssignmentInfo>> hostAssignments(Cluster cluster) {
+        Collection<SchedulerAssignment> assignmentValues =  cluster.getAssignments().values();
+        Map<String, List<AssignmentInfo>> hostAssignments = new HashMap<String, List<AssignmentInfo>>();
+
+        for (SchedulerAssignment sa : assignmentValues) {
+            Map<WorkerSlot, List<ExecutorDetails>> slotExecutors = Utils.reverseMap(sa.getExecutorToSlot());
+            Set<Map.Entry<WorkerSlot, List<ExecutorDetails>>> entries = slotExecutors.entrySet();
+            for (Map.Entry<WorkerSlot, List<ExecutorDetails>> entry : entries) {
+                WorkerSlot slot = entry.getKey();
+                List<ExecutorDetails> executors = entry.getValue();
+
+                String host = cluster.getHost(slot.getNodeId());
+                AssignmentInfo ass = new AssignmentInfo(slot, sa.getTopologyId(), new HashSet<ExecutorDetails>(executors));
+                List<AssignmentInfo> executorList = hostAssignments.get(host);
+                if (executorList == null) {
+                    executorList = new ArrayList<AssignmentInfo>();
+                    hostAssignments.put(host, executorList);
+                }
+                executorList.add(ass);
+            }
+        }
+        return hostAssignments;
+    }
+
+    private Set<Set<ExecutorDetails>> computeWorkerSpecs(TopologyDetails topology) {
+        Map<String, List<ExecutorDetails>> compExecutors = Utils.reverseMap(topology.getExecutorToComponent());
+
+        List<ExecutorDetails> allExecutors = new ArrayList<ExecutorDetails>();
+        Collection<List<ExecutorDetails>> values = compExecutors.values();
+        for (List<ExecutorDetails> eList : values) {
+            allExecutors.addAll(eList);
+        }
+
+        int numWorkers = topology.getNumWorkers();
+        int bucketIndex = 0;
+        Map<Integer, Set<ExecutorDetails>> bucketExecutors = new HashMap<Integer, Set<ExecutorDetails>>(numWorkers);
+        for (ExecutorDetails executor : allExecutors) {
+            Set<ExecutorDetails> executors = bucketExecutors.get(bucketIndex);
+            if (executors == null) {
+                executors = new HashSet<ExecutorDetails>();
+                bucketExecutors.put(bucketIndex, executors);
+            }
+            executors.add(executor);
+            bucketIndex = (++bucketIndex) % numWorkers;
```

Comment:
minor. since bucketIndex is being assigned anyway, ++bucketIndex can be replaced with (bucketIndex + 1)